webapp/graphite/storage.py

```diff
@@ -80,60 +84,60 @@ def find(self, pattern, startTime=None, endTime=None, local=False):
-          def distance_to_requested_interval(node):
-            latest = sorted(node.intervals, key=lambda i: i.end)[-1]
-            distance = query.interval.start - latest.end
-            return distance if distance >= 0 else float('inf')
-
-          best_candidate = min(leaf_nodes, key=distance_to_requested_interval)
-          if distance_to_requested_interval(best_candidate) <= settings.FIND_TOLERANCE:
-            minimal_node_set.add(best_candidate)
+      # Calculate best minimal node set
+      minimal_node_set = set()
+      covered_intervals = IntervalSet([])
+
+      # If the query doesn't fall entirely within the FIND_TOLERANCE window
+      # we disregard the window. This prevents unnecessary remote fetches
+      # caused when carbon's cache skews node.intervals, giving the appearance
+      # remote systems have data we don't have locally, which we probably do.
+      now = int( time.time() )
+      tolerance_window = now - settings.FIND_TOLERANCE
+      disregard_tolerance_window = query.interval.start < tolerance_window
+      prior_to_window = Interval( float('-inf'), tolerance_window )
+
+      def measure_of_added_coverage(node, drop_window=disregard_tolerance_window):
+        relevant_intervals = node.intervals.intersect_interval(query.interval)
+        if drop_window:
+          relevant_intervals = relevant_intervals.intersect_interval(prior_to_window)
+        return covered_intervals.union(relevant_intervals).size - covered_intervals.size
+
+      nodes_remaining = list(leaf_nodes)
+
+      # Prefer local nodes first (and do *not* drop the tolerance window)
+      for node in leaf_nodes:
+        if node.local and measure_of_added_coverage(node, False) > 0:
+          nodes_remaining.remove(node)
+          minimal_node_set.add(node)
+          covered_intervals = covered_intervals.union(node.intervals)
+
+      while nodes_remaining:
+        node_coverages = [ (measure_of_added_coverage(n), n) for n in nodes_remaining ]
+        best_coverage, best_node = max(node_coverages)
+
+        if best_coverage == 0:
+          break
+
+        nodes_remaining.remove(best_node)
+        minimal_node_set.add(best_node)
+        covered_intervals = covered_intervals.union(best_node.intervals)
+
+      # Sometimes the requested interval falls within the caching window.
+      # We include the most likely node if the gap is within tolerance.
+      if not minimal_node_set:
+        def distance_to_requested_interval(node):
+          latest = sorted(node.intervals, key=lambda i: i.end)[-1]
+          distance = query.interval.start - latest.end
+          return distance if distance >= 0 else float('inf')
+
+        best_candidate = min(leaf_nodes, key=distance_to_requested_interval)
+        if distance_to_requested_interval(best_candidate) <= settings.FIND_TOLERANCE:
+          minimal_node_set.add(best_candidate)
+
+      if settings.REMOTE_STORE_MERGE_RESULTS and not local and remote_nodes:
```

Comment:
Okay, the results are in:

* My comment about `datalib.py` can be ignored.  `storage.py` is the culprit alright.
* There *are* apparently overlaps with purely local nodes and [this code](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-57c0c4ffa18833a2e2ea55aab4309ffcR110) is helping to reduce that.
* I've included the tests for a possible fix as the last item.

### No slow down:
```diff
diff --git a/webapp/graphite/local_settings.py b/webapp/graphite/local_settings.py
--- a/webapp/graphite/local_settings.py
+++ b/webapp/graphite/local_settings.py
@@ -265,7 +265,7 @@ TIME_ZONE = 'Japan'
 #REMOTE_STORE_MERGE_RESULTS = True

 # 2016-09-02: this destroys performance
-REMOTE_STORE_MERGE_RESULTS = False
+#REMOTE_STORE_MERGE_RESULTS = False

diff --git a/webapp/graphite/storage.py b/webapp/graphite/storage.py
--- a/webapp/graphite/storage.py
+++ b/webapp/graphite/storage.py
@@ -80,7 +80,7 @@ class Store:
       if not leaf_nodes:
         continue

-      if settings.REMOTE_STORE_MERGE_RESULTS:
+      if False:
         # we need all nodes for merging results
         minimal_node_set = leaf_nodes
```

### Slowed down:

```diff
diff --git a/webapp/graphite/local_settings.py b/webapp/graphite/local_settings.py
--- a/webapp/graphite/local_settings.py
+++ b/webapp/graphite/local_settings.py
@@ -265,7 +265,7 @@ TIME_ZONE = 'Japan'
 #REMOTE_STORE_MERGE_RESULTS = True

 # 2016-09-02: this destroys performance
-REMOTE_STORE_MERGE_RESULTS = False
+#REMOTE_STORE_MERGE_RESULTS = False

diff --git a/webapp/graphite/storage.py b/webapp/graphite/storage.py
--- a/webapp/graphite/storage.py
+++ b/webapp/graphite/storage.py
@@ -80,7 +80,7 @@ class Store:
       if not leaf_nodes:
         continue

-      if settings.REMOTE_STORE_MERGE_RESULTS:
+      if True:
         # we need all nodes for merging results
         minimal_node_set = leaf_nodes
```

### Slowed down:

```diff
diff --git a/webapp/graphite/local_settings.py b/webapp/graphite/local_settings.py
--- a/webapp/graphite/local_settings.py
+++ b/webapp/graphite/local_settings.py
@@ -266,5 +266,5 @@ TIME_ZONE = 'Japan'

 # 2016-09-02: this destroys performance
-REMOTE_STORE_MERGE_RESULTS = False
+#REMOTE_STORE_MERGE_RESULTS = False

 ## Remote rendering settings
diff --git a/webapp/graphite/storage.py b/webapp/graphite/storage.py
--- a/webapp/graphite/storage.py
+++ b/webapp/graphite/storage.py
@@ -81,7 +81,7 @@ class Store:
         continue

-      if settings.REMOTE_STORE_MERGE_RESULTS:
+      if True:
         # we need all nodes for merging results
-        minimal_node_set = leaf_nodes
+        minimal_node_set = set(leaf_nodes)
       else:
         # Calculate best minimal node set
```

### No slow down: yay!
```diff
diff --git a/webapp/graphite/local_settings.py b/webapp/graphite/local_settings.py
--- a/webapp/graphite/local_settings.py
+++ b/webapp/graphite/local_settings.py
@@ -265,7 +265,7 @@ TIME_ZONE = 'Japan'
 #REMOTE_STORE_MERGE_RESULTS = True

 # 2016-09-02: this destroys performance
-REMOTE_STORE_MERGE_RESULTS = False
+#REMOTE_STORE_MERGE_RESULTS = False

 ## Remote rendering settings
 # Set to True to enable rendering of Graphs on a remote webapp
diff --git a/webapp/graphite/storage.py b/webapp/graphite/storage.py
--- a/webapp/graphite/storage.py
+++ b/webapp/graphite/storage.py
@@ -80,7 +80,7 @@ class Store:
       if not leaf_nodes:
         continue

-      if settings.REMOTE_STORE_MERGE_RESULTS:
+      if False:
         # we need all nodes for merging results
         minimal_node_set = leaf_nodes
       else:
@@ -112,16 +112,23 @@ class Store:
             minimal_node_set.add(node)
             covered_intervals = covered_intervals.union(node.intervals)

-        while nodes_remaining:
-          node_coverages = [ (measure_of_added_coverage(n), n) for n in nodes_remaining ]
-          best_coverage, best_node = max(node_coverages)
-
-          if best_coverage == 0:
-            break
-
-          nodes_remaining.remove(best_node)
-          minimal_node_set.add(best_node)
-          covered_intervals = covered_intervals.union(best_node.intervals)
+        if settings.REMOTE_STORE_MERGE_RESULTS:
+            for node in nodes_remaining:
+                if not node.local:
+                    nodes_remaining.remove(node)
+                    minimal_node_set.add(node)
+                    covered_intervals = covered_intervals.union(node.intervals)
+        else:
+            while nodes_remaining:
+              node_coverages = [ (measure_of_added_coverage(n), n) for n in nodes_remaining ]
+              best_coverage, best_node = max(node_coverages)
+
+              if best_coverage == 0:
+                   break
+
+              nodes_remaining.remove(best_node)
+              minimal_node_set.add(best_node)
+              covered_intervals = covered_intervals.union(best_node.intervals)

         # Sometimes the requested interval falls within the caching window.
         # We include the most likely node if the gap is within tolerance.
```